src/utils/price_comparison_utils.ts

```diff
@@ -118,8 +113,57 @@ export const priceComparisonUtils = {
                     gas = new BigNumber(gasScheduleWithOverrides[typedSource.liquiditySource]!(typedSource.fillData));
                 }
 
-                const unitMakerAmount = Web3Wrapper.toUnitAmount(makerAmount, buyToken.decimals);
-                const unitTakerAmount = Web3Wrapper.toUnitAmount(takerAmount, sellToken.decimals);
+                const unitMakerAmount = Web3Wrapper.toUnitAmount(source.makerAmount, buyToken.decimals);
+                const unitTakerAmount = Web3Wrapper.toUnitAmount(source.takerAmount, sellToken.decimals);
+
+                return {
+                    ...source,
+                    gas,
+                    unitTakerAmount,
+                    unitMakerAmount,
+                };
+            });
+
+            const ethUnitAmount = new BigNumber(10).pow(ethToken.decimals);
+            // NOTE: Sort sources by the best outcome for the user
+            // if the user is selling we want to maximize the maker amount they will receive
+            // if the user is buying we want to minimize the taker amount they have to pay
+            const sortedSources = isSelling
+                ? tradeSourcesWithGas.slice().sort((a, b) => {
+                      const aGasCostMakerAssetUnitAmount = a.gas
+                          .times(quote.gasPrice)
+                          .dividedBy(ethUnitAmount)
+                          .times(quote.ethToBuyTokenRate);
```

Comment:
Hey, should this be `dividedBy` instead of `times`?

Example:

1 eth = 3 buyTokens
ethToBuyTokenRate = 1:3 = 0.3333
gasInEth = 1
gasInTokens should be 3

**_with .times, wrong!_**
gasInTokens
= gasInEth * ethToBuyTokenRate
= 1 * 0.33333
= 0.333333 // wrong

**_with .dividedBy, correct!_**
gasInTokens
= gasInEth / ethToBuyTokenRate
= 1 / 0.33333
= 3 // expected

Reply:
I double checked in asset-swapper and I think the calculation here is correct, but the naming in asset-swapper is wrong, and this value actually represents XTokenToEthRate.

This is where the rate is sampled:
https://github.com/0xProject/protocol/blob/development/packages/asset-swapper/src/utils/market_operation_utils/index.ts#L166-L169

The comment here suggests it's "ETH -> maker token" ratio, e.g. 1 ETH to 4 tokens = 1:4 = 0.25

But the function that it's calling actually returns `(median maker token quote amount) / eth amount (1)`, i.e. 4 instead of 0.25:

https://github.com/0xProject/protocol/blob/development/packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts#L936

Reply:
Awesome find! Thanks for digging into it 🙏  The naming is definitely incorrect. I have now updated it to be `sellTokenToEthRate` and `buyTokenToEthRate`. Other than the naming I think the calculations here are correct but please let me know if you see any issue 🙏